src/types/res.rs

```diff
@@ -0,0 +1,140 @@
+use std::path::Path;
+
+use thiserror::Error;
+
+use crate::fnv1a64;
+use crate::raw::root::RED4ext as red;
+
+pub const MAX_LENGTH: usize = 216;
+
+#[derive(Debug, Default, Clone, Copy)]
+#[repr(transparent)]
+pub struct RaRef(red::RaRef);
+
+impl RaRef {
+    pub fn new(path: &(impl AsRef<Path> + ?Sized)) -> Result<Self, ResourcePathError> {
+        Ok(Self(red::RaRef {
+            path: red::ResourcePath {
+                hash: encode_path(path)?,
+            },
+        }))
+    }
+}
+
+#[derive(Debug, Default)]
+#[repr(transparent)]
+pub struct ResRef(red::ResRef);
+
+impl ResRef {
+    pub fn new(path: &(impl AsRef<Path> + ?Sized)) -> Result<Self, ResourcePathError> {
+        Ok(Self(red::ResRef {
+            resource: red::RaRef {
+                path: red::ResourcePath {
+                    hash: encode_path(path)?,
+                },
+            },
+        }))
+    }
+}
+
+impl Clone for ResRef {
+    fn clone(&self) -> Self {
+        Self(red::ResRef {
+            resource: self.0.resource,
+        })
+    }
+}
+
+fn encode_path(path: &(impl AsRef<Path> + ?Sized)) -> Result<u64, ResourcePathError> {
+    let sanitized = path
+        .as_ref()
+        .to_str()
+        .ok_or(ResourcePathError::InvalidUnicode)?;
+    let sanitized = sanitized
+        .trim_start_matches(['\'', '\"'])
+        .trim_end_matches(['\'', '\"'])
+        .trim_start_matches(['/', '\\'])
+        .trim_end_matches(['/', '\\'])
+        .split(['/', '\\'])
+        .filter(|comp| !comp.is_empty())
+        .map(str::to_ascii_lowercase)
+        .reduce(|mut acc, e| {
+            acc.push('\\');
+            acc.push_str(&e);
+            acc
+        })
+        .ok_or(ResourcePathError::Empty)?;
+    if sanitized.as_bytes().len() > self::MAX_LENGTH {
+        return Err(ResourcePathError::TooLong);
+    }
+    if Path::new(&sanitized)
+        .components()
+        .any(|x| !matches!(x, std::path::Component::Normal(_)))
+    {
+        return Err(ResourcePathError::NotCanonical);
+    }
+    Ok(fnv1a64(&sanitized))
+}
+
+#[derive(Debug, Error)]
+pub enum ResourcePathError {
+    #[error("resource path should not be empty")]
+    Empty,
+    #[error("resource path should be less than {} characters", self::MAX_LENGTH)]
+    TooLong,
+    #[error("resource path should be an absolute canonical path in an archive e.g. 'base\\mod\\character.ent'")]
+    NotCanonical,
+    #[error("resource path should be valid UTF-8")]
+    InvalidUnicode,
+}
+
+/// shortcut for ResRef creation.
+#[macro_export]
+macro_rules! res_ref {
+    ($base:expr, /$lit:literal $($tt:tt)*) => {
+        $crate::res_ref!([$base].join($lit), $($tt)*)
+    };
+    ($base:expr, ) => {
+        $base
+    };
+    ($lit:literal $($tt:tt)*) => {
+        $crate::types::ResRef::new(
+            &$crate::res_ref!($lit, $($tt)*).to_string()
+        )
+    };
+}
```

Comment:
this is calling `<[]>::join` which seems to be an incorrect way to implement this macro (the test doesn't check the created paths),
I'd fall back to the implementation in red4ext-rs
```suggestion
#[macro_export]
macro_rules! res_ref {
    ($base:expr, /$lit:literal $($tt:tt)*) => {
        $crate::res_ref!($base.join($lit), $($tt)*)
    };
    ($base:expr, ) => {
        $base
    };
    ($lit:literal $($tt:tt)*) => {
        $crate::ResRef::new(
            &$crate::res_ref!(::std::path::Path::new($lit), $($tt)*).to_string_lossy()
        )
    };
}
```

Reply:
adjusted in c9f1164, also completed test for created paths.